presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1116,34 +1116,76 @@ protected Scope visitJoin(Join node, Optional<Scope> scope)
             Scope left = process(node.getLeft(), scope);
             Scope right = process(node.getRight(), isLateralRelation(node.getRight()) ? Optional.of(left) : scope);
 
-            Scope output = createAndAssignScope(node, scope, left.getRelationType().joinWith(right.getRelationType()));
-
-            if (node.getType() == Join.Type.CROSS || node.getType() == Join.Type.IMPLICIT) {
-                return output;
-            }
-
             if (criteria instanceof JoinUsing) {
-                // TODO: implement proper "using" semantics with respect to output columns
                 List<Identifier> columns = ((JoinUsing) criteria).getColumns();
 
-                List<Expression> expressions = new ArrayList<>();
+                List<Field> joinFields = new ArrayList<>();
+
+                List<Integer> leftJoinFields = new ArrayList<>();
+                List<Integer> rightJoinFields = new ArrayList<>();
+
+                Set<Identifier> seen = new HashSet<>();
                 for (Identifier column : columns) {
-                    Expression leftExpression = new Identifier(column.getValue());
-                    Expression rightExpression = new Identifier(column.getValue());
+                    if (seen.contains(column)) {
+                        throw new SemanticException(DUPLICATE_COLUMN_NAME, column, "Column '%s' appears multiple times in USING clause", column.getValue());
+                    }
+                    seen.add(column);
+
+                    QualifiedName columnName = QualifiedName.of(column.getValue());
+                    Optional<ResolvedField> leftField = left.tryResolveField(column, columnName);
+                    Optional<ResolvedField> rightField = right.tryResolveField(column, columnName);
+
+                    if (!leftField.isPresent() || !rightField.isPresent()) {
+                        throw missingAttributeException(column, columnName);
+                    }
+
+                    try {
+                        metadata.getFunctionRegistry().resolveOperator(OperatorType.EQUAL, ImmutableList.of(
+                                leftField.get().getType(), rightField.get().getType()));
+                    }
+                    catch (OperatorNotFoundException e) {
+                        throw new SemanticException(TYPE_MISMATCH, node, "%s", e.getMessage());
+                    }
+
+                    Optional<Type> type = metadata.getTypeManager().getCommonSuperType(leftField.get().getType(), rightField.get().getType());
+                    analysis.addTypes(ImmutableMap.of(NodeRef.of(column), type.get()));
```

Comment:
Please make sure that `type` is present. Maybe `checkState` or `verify`

Reply:
By definition, it should exist, otherwise the equality operator above wouldn't be found. Optional.get() will fail as a last resort (if there's a bug).